api/app/Providers/AuthServiceProvider.php

```diff
@@ -65,41 +65,57 @@ public function boot(BearerTokenServiceInterface $tokenService)
+        });
+    }
 
-            $bearerToken = $request->bearerToken(); // 1. extract JWT access token from request.
-            if($bearerToken)
-            {
-                try {
-                    $claims = $tokenService->validateAndGetClaims($bearerToken);  // 2. validate access token.
-                    $sub = $claims->get('sub');
-                } catch (RequiredConstraintsViolated $e) {
-                    $violations = [];
-                    foreach ($e->violations() as $violationError) {
-                        array_push($violations, $violationError->getMessage());
-                    }
-                    Log::notice('Authorization token not valid: ', $violations);
-                    return abort(401, 'Authorization token not valid.');
-                } catch (Exception $e) {
-                    Log::notice('Error while validating authorization token: ' . $e->getMessage());
-                    return abort(401, 'Error while validating authorization token.');
+    /**
+     * Get the associated user model for a request or abort if something goes wrong with the lookup
+     *
+     */
+    public function resolveUserOrAbort($bearerToken, $tokenService): ?User {
+        if($bearerToken)
+        {
+            try {
+                $claims = $tokenService->validateAndGetClaims($bearerToken);  // 2. validate access token.
+                $sub = $claims->get('sub');
+            } catch (RequiredConstraintsViolated $e) {
+                $violations = [];
+                foreach ($e->violations() as $violationError) {
+                    array_push($violations, $violationError->getMessage());
                 }
+                Log::notice('Authorization token not valid: ', $violations);
+                return abort(401, 'Authorization token not valid.');
+            } catch (Exception $e) {
+                Log::notice('Error while validating authorization token: ' . $e->getMessage());
+                return abort(401, 'Error while validating authorization token.');
+            }
 
-                $userMatch = User::where('sub', $sub)->first(); // 3. match "sub" claim to user 'sub' field.
-                if($userMatch) {
-                    return $userMatch;
-                } else {
-                    Log::notice('No user found for given subscriber: '.$sub);
-                    return null;
-                }
+            // By this point we have verified that the token is legitimate
+
+            $userMatch = User::where('sub', $sub)->first(); // 3. match "sub" claim to user 'sub' field.
+            if($userMatch) {
+                return $userMatch;
+            } else {
+                // No user found for given subscriber - lets auto-register them
+                $newUser = new User;
+                $newUser->first_name = $sub;  // displayed on the landing page so should help us find the user
+                $newUser->email = $sub.'@new.user'; // junk, but email is currently non-nullable
```

Comment:
What do you think about changing our model so that email *is* nullable. It's not really a primary identifier anymore (that's sub, or id), its just contact info now. I like that if it is initialized as null, we can know to direct users to fill it in. 

We could add logic to migrations to disallow changing it *to* a null value, if we want.

Reply:
Yeah, definitely.  I'm on team "let's get rid of Laravel auth".  It doesn't really behave like SiC in a lot of ways and this is one of them. It's a very complicated project for what is essentially a poor mock or test double.  

I've added a migration to make email and sub both nullable.  I did sub since it's reasonable that many of our (non-admin) users will not have SiC subs to add.  I've tried to make the changes to the schema and app as light as possible for now.  I'll add an issue to the board to finish that job.

Reply:
#2722